Allow toggling touch input

NiLuJe opened the pull request:
The actual backend is not entirely too nasty, but it does require a nasty hack in Dispatcher for it to take immediately (otherwise, you need a restart). (Without it, you *can* disable input, but you *cannot* re-enable it ;p. It's not too bad because the onResume safety net does work, but it's still confusing).

Also includes a few cleanups/followups to #9717 ;).

(Rebase me).

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader/9738)
<!-- Reviewable:end -->

NiLuJe commented:
Draft because I'll be away for a few days, and in the hope that Dispatcher ninjas (ping @yparitcher) can think of a better way to deal with my quandary ;p.

NiLuJe commented:
(I went with monkey-patching because it means it doesn't affect  the code at all when unused/inactive, and gives me a de-facto way to know if it's currently enabled or not. Another approach would be to bake in the is_always_active guard in the common code-paths, paired with a class flag to enable filtering (i.e., two additional conditionals on every zone & ges_events)).

NiLuJe commented:
(Very quickly tested with a corner tap, I'm hoping it works everywhere, including multiswipes).

yparitcher commented:
Nasty hack not appreciated ;)

I pushed a alternate way (very lightly tested) to your branch (It was to annoying to try to explain) Feel free to nix it if necessary :)

NiLuJe commented:
Oh, neat :]. That's the spirit of what my idea looked like on paper before knowing how any of this actually worked, and realizing that InputContainer basically only had direct access to the zone's name, not its callback, which made me change tacks and tackle the issue the other way around ;). (Well, the that was the second hurdle, the first was realizing Gestures plugged its zones into the self.ui; not itself, the original plan went something along the lines of "cool, I'll just overload both InputContainer and Gesture's onGesture..." ;p).

I'll give it a try when I get back, thanks!

NiLuJe commented on plugins/gestures.koplugin/main.lua:
```diff
@@ -120,6 +121,12 @@ Multiswipes allow you to perform complex gestures built up out of multiple swipe
+-- If the gesture contains the "toggle_touch_input" action, mark it always_active
+-- to make sure that InputContainer won't block it after the IgnoreTouchInput Event.
+function Gestures:isGestureAlwaysActive(ges)
+    return self.gestures[ges] and self.gestures[ges].toggle_touch_input or nil
```
Note to self: the `or nil` is now redundant (I added it to clear the field instead of setting it to false; this is no longer a field ;)).

poire-z commented on frontend/ui/widget/container/inputcontainer.lua:
```diff
@@ -269,6 +270,58 @@ function InputContainer:onGesture(ev)
 
+-- Filtered variant that only lets is_always_active touch zones through.
+-- (This is used by the "toggle_touch_input" Dispatcher action).
+function InputContainer:_onGestureFiltered(ev)
+    for _, tzone in ipairs(self._ordered_touch_zones) do
+        if PluginShare.isGestureAlwaysActive(tzone.def.id) and tzone.gs_range:match(ev) and tzone.handler(ev) then
+            return true
```
Feels a bit awkward to have InputContainer dependant on this remote pluginshare module. (My preferred way for passing stuff around is to just store the thing in the ubiquitous UIManager without anybody knowing :)
(Also here, `PluginShare.isGestureAlwaysActive()` may crash if the gestures plugin is disabled and it is not set?)
Couldn't the gestures plugin just plug things in InputContainer ?

NiLuJe replied:
That looks sensible, yeah ;).

NiLuJe replied:
A fun quirk of the dummy approach here is that it probably shouldn't be a plain NOP, it might be safer to make it return true?

Frenzie commented on platform/kobo/koreader.sh:
```diff
@@ -611,10 +611,20 @@ if [ ${RETURN_VALUE} -ne ${KO_RC_HALT} ]; then
+        if pkill -0 kbmenu; then
+            # If we were started by KSM and requested an exit, do *NOT* exit the script so as not to re-enter KSM at all,
+            # to make sure the device powers off with our own ScreenSaver displayed.
+            while :; do
+                sleep 2073600
+            done
```
Er, this seems a bit excessive in every way imaginable?

Frenzie replied:
An endless while loop coupled with a sleep of hundreds of hours?! lol

Frenzie replied:
I'm thinking a simple 5-10 seconds (make it 30 or 60 if you like) without any while loops would do the trick. ;-)

Frenzie replied:
I know it's supposed to power off, but on the off chance it *doesn't* you're making it a hundred times harder to figure out that something went wrong, and worst case scenario it might drain the battery.

Frenzie replied:
Hence *not* doing that is significantly more harmless. There's no upside to value higher than 10s, imho.

NiLuJe replied:
Fair enough.

Frenzie commented on platform/kobo/koreader.sh:
```diff
@@ -611,10 +611,20 @@ if [ ${RETURN_VALUE} -ne ${KO_RC_HALT} ]; then
+        if pkill -0 kbmenu; then
+            # If we were started by KSM and requested an exit, do *NOT* exit the script so as not to re-enter KSM at all,
+            # to make sure the device powers off with our own ScreenSaver displayed.
+            while :; do
+                sleep 2073600
+            done
```
I.e.,
```suggestion
            sleep 10
```

NiLuJe replied:
This risks failing by exiting too early, or failing outright if interrupted by a signal ;).

NiLuJe replied:
The other approach would to just fudge the return code to 0, but that would still display KSM over our screensaver (I imagine?).

Frenzie replied:
The observed behavior is:

* KOReader shutdown screen
* KOReader shuts down and KSM takes over for a second

I don't see how adding 10 seconds risks exiting too early.

A signal will also shut down KSM.

Frenzie replied:
Forgot to add, I confirmed that it works as I said. I'm not talking in hypotheticals.

Frenzie commented on platform/kobo/koreader.sh:
```diff
@@ -611,10 +611,20 @@ if [ ${RETURN_VALUE} -ne ${KO_RC_HALT} ]; then
+            # so as not to re-enter KSM at all, to make sure the device powers off with our own ScreenSaver displayed.
+            # NOTE: This might not be fool-proof, as a poweroff might take longer than that,
+            #       or we might be interrupted early by signals.
+            sleep 30
```
That's an order of magnitude of a margin of error, but *soit* as they say in French. :-P

Frenzie replied:
Actually what I said isn't quite right, because after you shut the device down you expect to be able to start it back up again.

sleep 2 isn't quite enough to be 100% reliable; sleep 3 seems to do the trick reliably.

I'd round that up to 5 for safety and double it for additional safety. I'm not really sure where tripling it yet again comes into play. ^_^

NiLuJe commented:
I've had to tweak https://github.com/koreader/koreader/pull/9738/commits/5e2149aebed148d4460a1ab5d4b473588c828c1c to deal with multiswipes, and, once again, it's a bit clunky ;o).

Not sure if there's a better way, as I'm not entirely sure how they're actually registered? (ping @Frenzie and/or @yparitcher).

Frenzie commented:
Er, I don't really remember all the changes that were done. It was a hack that I took too far before @yparitcher so beautifully refactored it all, but I think the changes to the zones were actually done by yourself. (But even if it was the way I left it, I'd still have to read the code myself as well. ;-D)

poire-z commented on frontend/ui/widget/container/inputcontainer.lua:
```diff
@@ -269,6 +269,64 @@ function InputContainer:onGesture(ev)
     end
 end
 
+-- Will be overloaded by the Gestures plugin, if enabled, for use in _onGestureFiltered
+function InputContainer:isGestureAlwaysActive(ges, multiswipe_directions)
+    -- If the plugin isn't enabled, block everything, we'll restore standard behavior on resume.
+    return false
```
"block everything" ? shouldn't everything work if there is no way to restore the feature ?
"we'll restore standard behavior on resume." I'm a bit confused about thow this is related to suspend/resume. Or do you mean "on restore" ?

NiLuJe replied:
Resume restores the standard handler, so it makes sense to actually block every gesture, even without the Gesture plugin enabled. i.e., you could trigger this via Dispatcher somehow (Profiles? QuickMenu?), and just rely on the fact that once you suspend the device, everything will behave on resume as it restores gesture handling (this is apparently how this works on Kindles without buttons).

poire-z replied:
OK. Best to tell this somehow in the comment: "If the plugin isn't enabled, and invoked from Profiles or QuickMenu, still honor the request and block everything. The user can get out of it by suspending and resuming with the power button" (if I get what you mean right :) need to suspend before resuming. And this is not involved in a normal Suspend/Resume)

poire-z commented on frontend/ui/widget/container/inputcontainer.lua:
```diff
@@ -269,6 +269,64 @@ function InputContainer:onGesture(ev)
+    end
+end
+
+-- NOTE: This works because barely anything overloads onGesture.
+--       We would need to be slightly more creative if things did
+--       (i.e., don't stop the event and swap self.onGesture in every instance
+--       while still only swapping Input.onGesture once...)
+function InputContainer:onIgnoreTouchInput(toggle)
```
Might need a bit more context before your NOTE: :) at least saying what its "work" is :) and that it should be triggered by the "toggle_touch_input" gesture only.

poire-z commented on frontend/ui/widget/container/inputcontainer.lua:
```diff
@@ -269,6 +269,64 @@ function InputContainer:onGesture(ev)
+        end
+    elseif toggle == true then
+        -- Replace the onGesture handler w/ the minimal one if that's not already the case
+        if not InputContainer._onGesture then
+            InputContainer._onGesture = InputContainer.onGesture
+            InputContainer.onGesture = InputContainer._onGestureFiltered
```
I see we have (that you added probably) a lot of:
```
frontend/apps/reader/modules/readertoc.lua:function ReaderToc:onGesture() end
frontend/apps/reader/modules/readerconfig.lua:function ReaderConfig:onGesture() end
frontend/apps/reader/modules/readermenu.lua:function ReaderMenu:onGesture() end
frontend/apps/reader/modules/readerlink.lua:function ReaderLink:onGesture() end
frontend/apps/reader/modules/readerrotation.lua:function ReaderRotation:onGesture() end
frontend/apps/reader/modules/readerhighlight.lua:function ReaderHighlight:onGesture() end
frontend/apps/reader/modules/readerbookmark.lua:function ReaderBookmark:onGesture() end
frontend/apps/reader/modules/readerfont.lua:function ReaderFont:onGesture() end
frontend/apps/reader/modules/readerpaging.lua:function ReaderPaging:onGesture() end
frontend/apps/reader/modules/readerrolling.lua:function ReaderRolling:onGesture() end
```
How do these relate to what you are doing (you won't override these no-op which take precedence over your superclass monkey patch) ?

NiLuJe replied:
They aren't strictly related, hence the separate commit (or, actually, PR).

This is related to what #9710 unearthed, and was cleared up in 1e24a1c7a32f53329962ba503c969fcf690b800a

NiLuJe replied:
Basically, it's for ReaderUI modules that don't actually register their own gestures, but defer everything to ReaderUI.

Instead of having sendEvent go through a couple of empty arrays in the default InputContainer implementation, it just goes through a NOP implementation for those.

poire-z replied:
> it just goes through a NOP implementation for those

So, as it's a no-op, one that does not need to get overridden and replaced by your filtered one, right?

NiLuJe replied:
The very fact that it's implemented in a subclass means that we don't have anything extra to do (which is similar to what the comment is trying to convey), as we're only replacing the method of the base InputContainer class (which is also why we only do it once).

If we actually had meaningful ongesture subclass implementations[1], we'd need a broadcast event, and to replace self.onGesture, too.

[1] Notifications has one, but I figured dismissing notifs on tap wasn't really a hindrance ;).

poire-z commented on frontend/ui/widget/screensaverwidget.lua:
```diff
@@ -163,6 +165,9 @@ ScreenSaverWidget.onAnyKeyPressed = ScreenSaverWidget.onClose
 ScreenSaverWidget.onExitScreensaver = ScreenSaverWidget.onClose
 
 function ScreenSaverWidget:onCloseWidget()
+    Device.screen_saver_mode = false
+    Device.screen_saver_lock = false
```
Also a bit hard to grasp. They are plugged into Device (as some random holder of these flag), but not really documented, and their use is a bit scattered. Some single place documentation of these 2 flags could help.

NiLuJe replied:
They've indeed historically been scattered all over the place (but originated in Device) ;o).

Now, at least they're only *set* inside ScreenSaver/ScreenSaverWidget.

It kinda makes sense for them to be in Device, as it *is* a global state flag.

Il'l shove some comments about it... somewhere ;o).

poire-z commented on plugins/gestures.koplugin/main.lua:
```diff
@@ -120,6 +121,22 @@ Multiswipes allow you to perform complex gestures built up out of multiple swipe
+    --       Alas, all multiswipes are handled in a single multiswipe zone.
+    if self.multiswipes_enabled then
+        if ges == "multiswipe" and multiswipe_directions then
+            ges = "multiswipe_" .. self:safeMultiswipeName(multiswipe_directions)
```
(Probably peanuts vs the rest of things happening, but just asking: this (safeMultiswipeName() is done on each gesture, so not too costly ? although one gesture drawn per second is hard to do :)

NiLuJe replied:
It happens once per InputContainer, as long as it registers a `multiswipe` zone (`ges` is `multiswipe`), *and* the input event was actually a multiswipe (otherwise `multiswipe_directions` is `nil`).

(It's a single gsub call that replaces spaces with underscores, FWIW).

poire-z reviewed (commented):
May be it's just late or just me :) , but I had a hard time grasping it all -  dunno if the understanding I got after 20mn is right :/ More comments may help.

NiLuJe commented:
@poire-z: I *should* have addresses all your comments, ping me if I missed any or if something is still unclear ;).

Also tested the ScreenSaver stuff on the emu (doing Kindle now), and "fixed" the stupid-ass tests.

I'll fold in a fix for #9742 later tonight and this should be done ;).

NiLuJe reviewed (commented):
> Reviewed 2 of 6 files at r1, 1 of 5 files at r3, 4 of 4 files at r10, 10 of 10 files at r11, 7 of 7 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9738)* status: all files reviewed, 8 unresolved discussions (waiting on @Frenzie, @pazos, and @poire-z)



<!-- Sent from Reviewable.io -->

pazos reviewed (approved):
I didn't review really. Trusting you and @poire-z's review.

I'm here just to give you bonus points if you also remove https://github.com/koreader/koreader/blob/9863a9c0bb45487521b88be9d3278a77cb20bb91/frontend/device/input.lua#L573-L589 and https://github.com/koreader/koreader/blob/9bf19d1bb3b5c601383af71400003e71b7a78ed4/frontend/ui/elements/common_settings_menu_table.lua#L313-L318

Also remove the entry in menu orders. https://github.com/koreader/koreader/search?q=android_camera_key

If you want to skip all this I can take care of it once this is merged :+1: 

That was the crappy android specific implementation. I understand it is no longer needed.

:smile:

NiLuJe reviewed (commented):
> Reviewed 6 of 6 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9738)* status: all files reviewed, 8 unresolved discussions (waiting on @Frenzie and @poire-z)



<!-- Sent from Reviewable.io -->

poire-z commented:
>  Android: Get rid of the Camera key hack
Unnecessary now that we have a standard mechanism to disable touch input

Can one already associate Dispatcher actions to Android (and other devices) keys ?
Or will they have to use a touch gesture like the rest of us ?

NiLuJe commented:
Gestures, AFAIK. On the other hand, I imagine not many devices had a
dedicated camera key, so it's probably not a great loss ;).

On Tue, Nov 8, 2022, 19:35 poire-z ***@***.***> wrote:

> Android: Get rid of the Camera key hack
> Unnecessary now that we have a standard mechanism to disable touch input
>
> Can one already associate Dispatcher actions to Android (and other
> devices) keys ?
> Or will they have to use a touch gesture like the rest of us ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/koreader/koreader/pull/9738#issuecomment-1307663483>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA3KZRLCAVBNMO5LMNWXGLWHKMNRANCNFSM6AAAAAARVSZT4Y>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

hius07 commented:
A user can be trapped if toggles touch input via a profile without assigning a gesture.

Frenzie commented:
Are you still trapped if you do a standby/wake up with the button?

hius07 commented:
Great, have missed this!

Frenzie commented:
It's a good remark regardless, but then at least there's a workaround in the meantime.

NiLuJe commented:
I was just thinking about how to document that...

@hius07/@yparitcher: Can Dispatcher actions have a long-press "help" message?

hius07 commented:
> Can Dispatcher actions have a long-press "help" message?

Long-press is free for this category of actions but is used to clear (uncheck) the action for other categories, so no.

poire-z commented:
Using some old version of KSM, and now, when just "Exit koreader" (to get to KSM and enable usb, etc...), the device reboots ! _Sad!_
```
11/12/22-13:46:08 INFO  no dialog left to show
11/12/22-13:46:08 INFO  quitting uimanager with exit code: 0
```

```
/mnt/onboard/.adds/koreader # ps auxw|grep kbmenu
  654 root       0:00 {ksmhome.sh} /bin/sh /mnt/onboard/.adds/kbmenu/onstart/ksmhome.sh
 1053 root       0:00 {start_koreader.} /bin/sh /mnt/onboard/.adds/kbmenu/onstart/start_koreader.sh
 1134 root       0:00 grep kbmenu

/mnt/onboard/.adds/koreader # pgrep -f kbmenu && echo ok
654
1053
ok

/mnt/onboard/.adds/koreader # pkill -0 kbmenu && echo ok
/mnt/onboard/.adds/koreader #
```

Frenzie commented:
I can confirm, something changed since the version I tested before.

Frenzie commented:
Actually come to think of it, I only tested poweroff. There's probably a simple logic mistake.

poire-z commented:
I didn't follow the logic and the need for a change, but it feels it's just the switch to using `pkill -0` instead of `pgrep -f` that causes the issue.

Frenzie commented:
The change is because power off was broken at some point. I have no idea when because I didn't even realize that code path was touched.

Frenzie commented:
Ohh I see what you mean now. I don't think I tested *that*; that's a random @NiLuJe tack on xp

NiLuJe commented:
Okay, I was mildly afraid of that ;o).

Just adding `-f` to the pkill stuff will fix it, i'll do that.
